### ISSUE TITLE
Add ability to sort tasks

### ISSUE THREAD

**Issue by KenBW2** [owncloud/tasks#86]
At the moment the only way to sort tasks in the UI is to add sequential to-do dates to them, but this kind of abuses the system for non-date tasks.

It'd be good to be able to drag-sort tasks. I gather in the underlying storage backend this would mean reordering the tasks

**Comment by DemoFreak** [comment]
+1
Sorting tasks by priority (-2 to +2 would be sufficient, and would be the best choice due to syncing to mobile apps) or by dragging them around is IMHO an important feature.

**Comment by KenBW2** [comment]
I think it's important to stay true to the underlying ical format, so let's not bake in Owncloud-specific features.

The only way I can see (unless anyone knows a better way) is to just adjust the order they're stacked in in the database and/or the ics file

**Comment by DemoFreak** [comment]
The ical standard defines the PRIORITY property for VTODO events, so why not using this? The Tasks app already makes use of it, the only thing to do is a GUI improvement (so far as I understand it).

**Comment by raimund-schluessler** [comment]
Yes, the app does use the priority property for marking tasks as important, but there is no UI to use other priority values than 1 (highest priority) or 0 (undefined priority). Other valid values would be in the range [0 .. 9]. However, if two tasks have the same priority, you would not be able to sort them. Hence, in case you have more than 9 tasks there would always be an ambiguity when sorting tasks only by priority (which therefore is a separate issue https://github.com/owncloud/tasks/issues/77)

Here you would not be able to sort the tasks a, b, c, although the groups (abc), (d), (e) would be sorted, which may be a first step.
--- priority 9 ---
task a
task b
task c
--- priority 8 --
task d
--- priority 7 --
task e

**Comment by dobek** [comment]
It'd better then the current.

**Comment by DemoFreak** [comment]
> However, if two tasks have the same priority, you would not be able to sort them. Hence, in case you have more than 9 tasks there would always be an ambiguity when sorting tasks only by priority (which therefore is a separate issue #77)

That's - IMHO - no problem at all as sorting is mostly meant for having the most important tasks on top of the list. If several tasks are equally important then their inner order is rather irrelevant. In this context it would be nice if there would be a separating line or sth else between the priorities.

**Comment by raimund-schluessler** [comment]
Well, the most important tasks are already on top, when you click on the star.

**Comment by DemoFreak** [comment]
Of course, but you ain't able to prioritize more than that.
"First of all you have to do A and C, then B, E and F, and after that there is D, Y and Z" ;-)
Outlook has "low - medium - high", KDEs PIM allows 0 (no prio) to +9 (highest prio).
As Owncloud syncs to both of them it would be nice to have this option in Ownclouds builtin app too.

**Comment by DemoFreak** [comment]
> Outlook has "low - medium - high", KDEs PIM allows 0 (no prio) to +9 (highest prio).

Silly me, it's actually inverted.
0 (no prio), 9 (lowest) to 1 (highest).

**Comment by DemoFreak** [comment]
As another approach, maybe it'd be easier to simply sort the tasks alphabetically.
So you were able to sort equally to the most desktop applications, as you can sort by a simple click to the column header there, usually.

**Comment by dobek** [comment]
It's smart trick but it would be difficult to implement changing the priority by drag and drop. I believe it's what we finally want.

However, We could add button to renumber e.g.
before:
0. urgent 
3. task a
3.1 task c
5. task b
after:
1. urgent
2. task a
3. task c
4. task b

Drag and drop could renumber all tasks matching the pattern "[?.]+ text"

We could even think about subtasks e.g.

1. a
2.1. b
2.1;1. c
2.1;2. d

**Comment by DemoFreak** [comment]
Ok, I'm fine with each approach, as long as the order is synced to CALDAV clients in any manner.
Renaming would be synced, as like as PRIORITY would. Ordering only by drag'n'drop is fine for web view, but it won't sync. ;-)

**Comment by bennidi** [comment]
First of all +1 for the sorting feature. I think it is a very important feature in task management. After all, you always end up with a lot of different tasks and you need to prioritize. Dual classification (importan/not important) is not enough.

The PRIORITY field of the ical spec does not seem very appropriate as it supports only a very limited range of values.
For arbitrary sorting each calendar would need its own sequence that is always incremented when a new task is created. Sorting would then only involve swapping of tasks priority index.

Implementation could be done by defining a custom field or hijacking an exotic field to use it for arbitrary meta data. I assume that there is a way to make parsers ignore unknown data. Such a mechanism would allow to inject any kind of meta data into a task, hence making many other features possible.

Another approach would be to separate the metadata from the calendar entries and use a task id to fetch metadata. Probably a good way for all data that does not need to be synced to other iCal clients. IMHO task priorities do not need to by synced but I assume many others would disagree on that matter.

Anyways, thanks for that great app. It is really helpful and has a good look and feel. Keep up the good work!

**Comment by vjanuario** [comment]
When syncing to iOS via CalDAV, "X-APPLE-SORT-ORDER" is added to the tasks and saved in the database with the task. Implementing sort with this variable would have the advantage that sorting on ownCloud would mirror that done on Apple devices. ownCloud already has drag-n-drop to move tasks into lists. I know it's much easier said than done and I appreciate the effort the developers have put into creating ownCloud. All the pieces are there if someone has time to put it together.

**Comment by raimund-schluessler** [comment]
I will implement this feature next, but I don't really get how Apple uses the ```X-APPLE-SORT-ORDER``` property. When I create 5 tasks (either using iOS or this app here), none of it has the ```X-APPLE-SORT-ORDER``` property. When I change the position of task 3, then only task 3 gets this property set to a seemingly arbitrary number. So there must be some sort of mechanism which calculates a value for the other tasks to which ```X-APPLE-SORT-ORDER``` is compared in order to sort correctly. Could someone give me any insight to this?

**Comment by raimund-schluessler** [comment]
With #357 the task list can now be sorted by the default order (duedate, priority, start, summary), by due date, by start date, by priority and alphabetically.

**Comment by DemoFreak** [comment]
Sorting todos in the web GUI works like a charm. :+1: 
(Tested on NC 10.0)

Is it planned to implement syncing the sort order to CalDAV clients?

**Comment by raimund-schluessler** [comment]
It depends on the clients if they are able to sort the tasks accordingly. All information necessary to sort the tasks will be present in the VTODO.

**Comment by raimund-schluessler** [comment]
> So there must be some sort of mechanism which calculates a value for the other tasks to which X-APPLE-SORT-ORDER is compared in order to sort correctly.

Note to myself: The "arbitrary number" apple uses as value for ```X-APPLE-SORT-ORDER``` is the number of seconds between the creation of a task and ```20010101T000000Z```.

**Comment by KenBW2** [comment]
FWIW I prefer the standard VTODO approaches to this. As much as it might be tempting to come up with a custom sort order, it's of no use if this can't reliably be synced to CalDAV clients

**Comment by raimund-schluessler** [comment]
This is a "standard VTODO approach" and the sort order is synced to the client. ```X-APPLE-SORT-ORDER``` is present in the ```VTODO``` and could be evaluated by every client. However, we can't force a client to evaluate a certain sort order. If the client does not support sorting tasks, there is nothing we can do about.

We can also implement further sorting keys other than ```X-APPLE-SORT-ORDER``` but I am not aware of any other sorting key, vendor specific or not.

**Comment by edmondscommerce** [comment]
I agree it's up to the client to do that actual sorting. But I'd be hesitant to bake into ownCloud the expectation of nonstandard/vendorspecific extensions to VTODO - namely the `X-APPLE-SORT-ORDER` one.

I know standard VTODO doesn't have a sort order. The only way I could see it being implemented without these is to reorder them in the backend (database) although I realise that's far from reliable.

The pull request #357 as it stands is good - it's just changing the view on the standard data, which is what a CalDAV client *should* do

**Comment by KenBW2** [comment]
I agree it's up to the client to do that actual sorting. But I'd be hesitant to bake into ownCloud the expectation of nonstandard/vendorspecific extensions to VTODO - namely the `X-APPLE-SORT-ORDER` one.

I know standard VTODO doesn't have a sort order. The only way I could see it being implemented without these is to reorder them in the backend (database) although I realise that's far from reliable.

The pull request #357 as it stands is good - it's just changing the view on the standard data, which is what a CalDAV client *should* do

**Comment by raimund-schluessler** [comment]
We just use `X-APPLE-SORT-ORDER` because it is the only sort extension I am aware of. And since there is no standard solution we have to come up with something ourselves. We could also use our own key like `X-OC-SORT-ORDER`, but the benefit of `X-APPLE-SORT-ORDER` is the compatibility with Apple. If there are other extensions used for sorting, we could implement these too and sync them with `X-APPLE-SORT-ORDER`.

Also, it is not a problem, if `X-APPLE-SORT-ORDER`  is not present in a VTODO, because then (if the manual sort order is chosen) we just order it by the created date (which is basically what Apple does too, for tasks which were created by other clients). So, I don't see any drawback if we use `X-APPLE-SORT-ORDER`. Worst thing that could happen is that an other client does not support sorting by this key, but this is no worse than not having this key at all. Also #357 is not affected by this, we just introduce an other option for sorting tasks manually.

So basically, I agree with you. The client is responsible for sorting. But since there is no standard sort option, we could either not have a custom sort at all, or implement our own solution, which allows sorting in the webview and also syncs with Apple devices.
